moby/moby
# [WIP] docker inspection fs (/dev/docker in a container), for exposing the Swarm task slot

**\- What I did**

Provides the following files for exposing task information (especially, the task slot) to containers:
- `/dev/docker/container/json`
- `/dev/docker/swarm/task/json`

`/dev/docker` is mounted by default for tasks, but not mounted for non-task containers. (you can enable it by `docker run --inspectionfs`)

For the story so far, please look into https://github.com/docker/docker/issues/24110#issuecomment-234179743

ping @stevvooe @vbatts @SvenDowideit

Fix https://github.com/docker/docker/issues/24110
Update https://github.com/docker/docker/issues/23843
Update https://github.com/docker/docker/issues/24113

**\- How I did it**
Used FUSE.

**\- How to verify it**
- Create a service: `docker service create --replicas 3 --name s1 alpine tail -f /dev/null`
- Do `cat /dev/docker/container/json` in each of task containers
- Do `cat /dev/docker/swarm/task/json` in each of task containers
- Remove the service: `docker service rm s1`
- Make sure the FUSE filesystem are unmounted: `cat /proc/mounts`

---

Discussion:
- Is it secure?
- Is it machine-friendly?
- Is it human-friendly?
- Alternative ideas for passing per-task info:
  - @stevvooe 's `docker service create -e ZKID={{.Task.Slot}}` https://github.com/docker/docker/issues/24110#issuecomment-232848838
  - my previous proposal:  `docker service create --scatter=ZKID=$(seq 1 3)` https://github.com/docker/docker/issues/23843
  - Expose REST socket to containers (I don't like it)

TODOs:
- Hide `/dev/docker/swarm` if the container is not a swarm task
- Separete modifications under `vendor`
- Fix compilation errors for Windows
